### ISSUE TITLE
关于 batch size 和 training steps

### ISSUE THREAD

**Issue by Leon-0602** [lzhxmu/CPPO#5]
感谢 CPPO 作者的开源，很有趣的一份工作~

在论文的 Figure 4 中，我看到 GRPO 在 GSM8k 和 MATH 数据集上的步数分别为大约 7000 和大约 2500.

请问这里的 steps 数是怎么得到的？比如，每个 step 对应着多少次参数更新/多少个 questions/多少个 responses 呢？

我按照论文里写的 batch size = 16, group size=16, gpu_num=2，结合 GSM8k 和 MATH 的样本个数，无论如何也无法理解为什么 GRPO 的步数分别是大约 7000 和 2500.

能否烦请作者详细解释一下？

感谢。

**Comment by lzhxmu** [comment]
感谢你关注我们的工作，在 openr1 框架中 gpu_num=2，其实有一张是用于 vllm 进行 Completions 生成的，所以实际上等效于一个进程在训练，batch size=16 是指一次前向有 16 个 response/Completions 参与参数更新， 在 batch size = 16, group size=16, gpu_num=2 设置时，实际上一次前向就是一个 question的 16 个 responses， GSM8K 有 7k 左右数据，因此steps就是 7k 左右，math 用的 4 卡，所以等效于 3 个进程并行，就是question 数量除 3。

**Comment by Leon-0602** [comment]
感谢您的超快回复！我大概明白了，“vllm 会单独占用一张 GPU 用于回复生成” 这个点我之前没意识到。

然而，我看您提供的脚本和配置文件中，num_iterations = 2，也就是说**所有问题会被复用一次，那么 training_steps 应该会乘2吧？**

我用默认超参跑您的代码，看到的 training_steps 数也是乘了2的（一共14946步）：

![Image](https://github.com/user-attachments/assets/1f0a4c17-ee03-4d2f-8834-1abc1ecc704f)

**Comment by lzhxmu** [comment]
> 感谢您的超快回复！我大概明白了，“vllm 会单独占用一张 GPU 用于回复生成” 这个点我之前没意识到。
> 
> 然而，我看您提供的脚本和配置文件中，num_iterations = 2，也就是说**所有问题会被复用一次，那么 training_steps 应该会乘2吧？**
> 
> 我用默认超参跑您的代码，看到的 training_steps 数也是乘了2的（一共14946步）：
> 
> ![Image](https://github.com/user-attachments/assets/1f0a4c17-ee03-4d2f-8834-1abc1ecc704f)

是所有问题的Completions都会被复用一次，Completions不变就说明 reward 没变，所以横坐标就是 7k 多，如果算梯度更新次数的话，确实要乘上num_iterations～

**Comment by Leon-0602** [comment]
再次向您表示感谢！但是抱歉，还有一个重要问题，是关于 batch size 选取的：

目前一些热门的 RL 框架（例如 veRL）中支持非常大的 global_batch_size（即多少个 questions 放在一起生成 responses）和 mini_batch_size（即多少个 question-response pair 放在一起更新一次模型参数）。比如，global_batch_size=512, mini_batch_size=512, group_size = 16 就意味着 512 个 questions 会用同一个 step 下的模型参数生成 512 * 16 个 responses，然后这 512 * 16 个 question-response pairs 会分成16份（每一份包含 512 个 question-response pairs）来更新模型参数。这样算下来，相当于【16次参数更新】对应【512个questions】，即【1次参数更新】对应【32个questions】。

这样的 batch size 设置与 CPPO 的设定有至少两点不同：

1. batch size 相差过大。从“1次参数更新对应多少 questions” 的角度看，
    - CPPO 在 gpu_num=2 时的这个数字是1
    - 然而， veRL 中的这个数字是32

2. 参数迭代方式不同。
    - CPPO 中是【参数k生成回答】->【参数k计算loss并更新参数】->【参数k+1生成回答】->【参数k+1计算loss并更新参数】->...
    - 然而，veRL 中是【参数k生成回答】->【参数 k 计算 loss 并更新参数】->【参数 k+1 计算 loss 并更新参数】->....->【参数 k+15 计算 loss 并更新参数】->【参数 k+16 生成回答】->...

请问您如何看待这一 batch size 设定上的极大差异？

**Comment by lzhxmu** [comment]
这个是OpenR1 的实现没有 verl 高效，所以导致 batch size 开不大，会爆显存，openr1 的框架有待进一步优化，batch size 会影响训练结果，但是只要学习率调好，应该不会相差太大，verl 的实现方式有点像 off-policy 了，因为模型更新了 n 次，但是还是用 n 次之前的 response 来训，我个人认为是一种近似，对结果的影响还有待验证。本文只是选取 openr1 来进行验证想法，理论上用 verl 也是可行的，需要做实验进一步验证。

**Comment by Leon-0602** [comment]
> 这个是OpenR1 的实现没有 verl 高效，所以导致 batch size 开不大，会爆显存，openr1 的框架有待进一步优化，batch size 会影响训练结果，但是只要学习率调好，应该不会相差太大，verl 的实现方式有点像 off-policy 了，因为模型更新了 n 次，但是还是用 n 次之前的 response 来训，我个人认为是一种近似，对结果的影响还有待验证。本文只是选取 openr1 来进行验证想法，理论上用 verl 也是可行的，需要做实验进一步验证。

好的，谢谢您的耐心解答！祝您工作顺利，生活愉快。
